### PR TITLE
ColumnDataSource.from_dataframe using arrays

### LINKED ISSUE #5973: CDS creation from DataFrames should not use tolist()
Discovered fixing #5152: Currently Pandas DataFrames still convert all their columns to python lists using `.tolist` Not only is this slow, it means that no CDS creates this way will ever make use of the binary array protocol. One option is to use:
```
ds.to_dict('series')
```
which returns a "dict of series" however the series objects are references the actual same series objects in the dataframe. Not sure if sharing is advised. Copying to equivalent numpy array seems advised.

### PR DESCRIPTION
**Pull request by jsignell**
- [x] issues: fixes #5973

### CODE REVIEW

File: bokeh/util/serialization.py [hunk 1]
```diff
@@ -150,6 +150,10 @@ def transform_array_to_list(array):
         transformed[np.isposinf(array)] = 'Infinity'
         transformed[np.isneginf(array)] = '-Infinity'
         return transformed.tolist()
+    elif (array.dtype.kind == 'O' and pd.isnull(array).any()):
```
Comment (bryevdv) [inline bokeh/util/serialization.py]: we don't assume `pandas` as a hard dependency, `pd` can be none if `optional_import` above fails, so will need to check that.

Comment (jsignell) [inline bokeh/util/serialization.py]: It it better to look for matches to np.nan? That seems dicey. And np.isnan doesn't work

Comment (bryevdv) [inline bokeh/util/serialization.py]: if the problem is that there are actual pandas nulls then the solution is just to and `and pd` to the if above I think? Can't have pandas nulls without pandas present.

Comment (jsignell) [inline bokeh/util/serialization.py]: you can have an object array with nan in it even if pandas isn't there though right?

Comment (bryevdv) [inline bokeh/util/serialization.py]: certainly but they won't have pandas nulls in them without pandas installed :)

Comment (jsignell) [inline bokeh/util/serialization.py]: This is what I am worried about:

```python
>>> a = np.array(['a', 'b' ,4, np.nan])
>>> np.isnan(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
>>> pd.isnull(a)
array([False, False, False, False], dtype=bool)
```

Comment (jsignell) [inline bokeh/util/serialization.py]: oh wait I read that wrong

File: bokeh/models/sources.py [hunk 2]
```diff
@@ -117,19 +117,20 @@ def _data_from_df(df):
...
+        _df = df.copy()
+        index = _df.index
+        new_data = _df.to_dict('series')
+
         if index.name:
-            new_data[index.name] = index.tolist()
+            new_data[index.name] = index.values
         elif index.names and not all([x is None for x in index.names]):
-            new_data["_".join(index.names)] = index.tolist()
+            print("_".join(index.names))
```
Comment (bryevdv) [inline bokeh/models/sources.py]: `print` should be removed

File: bokeh/models/sources.py [hunk 3]
```diff
@@ -141,7 +142,7 @@ def from_df(cls, data):
             data (DataFrame) : data to convert
 
         Returns:
-            dict[str, list]
+            dict[str, np.array]
```
Comment (bryevdv) [inline bokeh/models/sources.py]: I think this will always return a `pd.Series`? We can be specific if so

File: bokeh/util/serialization.py [hunk 4]
```diff
@@ -150,6 +150,10 @@ def transform_array_to_list(array):
         transformed[np.isposinf(array)] = 'Infinity'
         transformed[np.isneginf(array)] = '-Infinity'
         return transformed.tolist()
+    elif (array.dtype.kind == 'O' and pd and pd.isnull(array).any()):
+        transformed = array.astype('object')
+        transformed[pd.isnull(array)] = 'NaN'
+        return transformed.tolist()
```
Comment (bryevdv) [inline bokeh/util/serialization.py]: We need a test for this.

File: bokeh/models/sources.py [hunk 5]
```diff
@@ -141,7 +142,7 @@ def from_df(cls, data):
...
 
         '''
         return cls._data_from_df(data)
```
Comment (bryevdv) [inline bokeh/models/sources.py]: I'd suggest a couple of tests, at a minimum we should maintain that passing a DataFrame results in a `.data` with `pd.Series` as column values

File: bokeh/models/sources.py (hunk 3 above)
Comment (jsignell) [inline bokeh/models/sources.py]: the index is an `np.array`

**Comment by bryevdv** [comment]
Test failures are purely due to intersphinx faillures caused by `scipy.org` being down. Going to go ahead and merge now.

**Pull request merged**